## TODO: DELETE THIS NOTEBOOK BEFORE SUBMISSION
# Milestone 2: Insight Of Data And Feasibility Check Of Our Idea

<hr style="clear:both">

This notebook traces our journey through the feasibility check of our idea : Checking trends in music composition for the movie industry.

**Project Mentor:** [Aoxiang Fan](https://people.epfl.ch/aoxiang.fan) ([Email](mailto:aoxiang.fan@epfl.ch))
**Authors:** [Luca Carroz](https://people.epfl.ch/emilie.carroz), [David Schroeter](https://people.epfl.ch/david.schroeter), [Xavier Ogay](https://people.epfl.ch/xavier.ogay), [Joris Monnet](https://people.epfl.ch/joris.monnet), [Paulo Ribeiro de Carvalho](https://people.epfl.ch/paulo.ribeirodecarvalho)

<hr style="clear:both">

## Storyline

**Title : Stanislas' music dream : Road to Hollywood !**

A 20-year-old aspiring musician, Stanislas, fueled by a passion for the film industry, embarks on a quest to launch his career. His ultimate dream? To hear one of his productions featured in a Hollywood film and become one of the planet's top composers. To increase his chances, he turns to a team of Data Scientists known as LSD.

The "LearningtheSecretsofData" team's mission is to identify trends shared among successful music composers and compositions, ultimately optimizing choices for our young musician. This is not an easy task but the team is driven by the wish of helping Stanislas. How could they provoke a cascADA of successful choices in Stany career.

Which music genre Stany should he focus on? Will this new direction be enough for him to conquers the show business? Maybe he may invest in a ludicrous website to promote himself? Or should he even consider changing Nationality to achieve his goal? Let’s see what’s the plan LSD had concocted for Stanislas.


## Needed data

To help Stanislas with his dream, we will need some specific information about each analysed movie. To support our final answer concerning potential trends in composer for movie industry, we will analyse few questions :

1) Which are the most frequent music genre appearing in movies ?
2) What is the average composer's age at their :
   - first movie appearance ?
   - biggest box office revenue ?
3) How the top composers' career progress over the years ?
4) Where do composers come from ?
5) Does composer's gender matter ?
6) Does having a personal website correlate with the composers' success ?
7) Is there a correlation between box office revenue and movie's playlist popularity ?

List of needed information about movie's **composer** :

- Name
- Birthday
- Gender
- Homepage
- Place of birth
- First appearance in movie credits

List of needed information about composers' **musics** :

- Genre
- Spotify's popularity


## Import

In [1]:
# Import all needed libraries
from helpers import *

# Load autoreload extension
%load_ext autoreload

# Set autoreload mode
%autoreload 2

## Load Data

Only `movie.metadata.tsv` is loaded, since others data sets are not relevant for our data analysis.

Note that no header are given in the raw data sets. We then looked at the documentation and set these headers with meaningful names. Please, find the documentation of data set [here](http://www.cs.cmu.edu/~ark/personas/).

In [2]:
# Please be sure your data is store in the same path
data_path = 'dataset/MovieSummaries/movie.metadata.tsv'

# Load the text file into Pandas DataFrame
movie = load_movies(data_path)

# Display the dataframe
display(movie)

C:\Users\41786\OneDrive\Documents\EPFL\master\ma1\ADA\ada-2023-project-learningthesecretsofdata\venv\lib\site-packages\pandas\core\dtypes\cast.py:1056: RuntimeWarning: invalid value encountered in cast
  if (arr.astype(int) == arr).all():
C:\Users\41786\OneDrive\Documents\EPFL\master\ma1\ADA\ada-2023-project-learningthesecretsofdata\venv\lib\site-packages\pandas\core\dtypes\cast.py:1080: RuntimeWarning: invalid value encountered in cast
  if (arr.astype(int) == arr).all():


wiki_movieID freebase_movieID  \
0            975900        /m/03vyhn   
1           3196793        /m/08yl5d   
2          28463795       /m/0crgdbh   
3           9363483       /m/0285_cd   
4            261236        /m/01mrr1   
...             ...              ...   
81736      35228177       /m/0j7hxnt   
81737      34980460       /m/0g4pl34   
81738       9971909       /m/02pygw1   
81739        913762        /m/03pcrp   
81740      12476867       /m/02w7zz8   

                                                    name release_date  \
0                                         Ghosts of Mars   2001-08-24   
1      Getting Away with Murder: The JonBenét Ramsey ...   2000-02-16   
2                                            Brun bitter         1988   
3                                       White Of The Eye         1987   
4                                      A Woman in Flames         1983   
...                                                  ...          ...   
81736                           Mermaids: The Body Found   2011-03-19   
81737                                            Knuckle   2011-01-21   
81738                                  Another Nice Mess   1972-09-22   
81739  The Super Dimension Fortress Macross II: Lover...   1992-05-21   
81740                                            Spliced         2002   

       box_office_revenue  runtime                           languages  \
0              14010832.0     98.0  {"/m/02h40lc": "English Language"}   
1                    <NA>     95.0  {"/m/02h40lc": "English Language"}   
2                    <NA>     83.0  {"/m/05f_3": "Norwegian Language"}   
3                    <NA>    110.0  {"/m/02h40lc": "English Language"}   
4                    <NA>    106.0   {"/m/04306rv": "German Language"}   
...                   ...      ...                                 ...   
81736                <NA>    120.0  {"/m/02h40lc": "English Language"}   
81737                <NA>     96.0  {"/m/02h40lc": "English Language"}   
81738                <NA>     66.0  {"/m/02h40lc": "English Language"}   
81739                <NA>    150.0   {"/m/03_9r": "Japanese Language"}   
81740                <NA>     86.0  {"/m/02h40lc": "English Language"}   

                                               countries  \
0              {"/m/09c7w0": "United States of America"}   
1              {"/m/09c7w0": "United States of America"}   
2                                 {"/m/05b4w": "Norway"}   
3                         {"/m/07ssc": "United Kingdom"}   
4                                {"/m/0345h": "Germany"}   
...                                                  ...   
81736          {"/m/09c7w0": "United States of America"}   
81737  {"/m/03rt9": "Ireland", "/m/07ssc": "United Ki...   
81738          {"/m/09c7w0": "United States of America"}   
81739                              {"/m/03_3d": "Japan"}   
81740                            {"/m/0d060g": "Canada"}   

                                                  genres  
0      {"/m/01jfsb": "Thriller", "/m/06n90": "Science...  
1      {"/m/02n4kr": "Mystery", "/m/03bxz7": "Biograp...  
2      {"/m/0lsxr": "Crime Fiction", "/m/07s9rl0": "D...  
3      {"/m/01jfsb": "Thriller", "/m/0glj9q": "Erotic...  
4                                {"/m/07s9rl0": "Drama"}  
...                                                  ...  
81736                            {"/m/07s9rl0": "Drama"}  
81737  {"/m/03bxz7": "Biographical film", "/m/07s9rl0...  
81738       {"/m/06nbt": "Satire", "/m/01z4y": "Comedy"}  
81739  {"/m/06n90": "Science Fiction", "/m/0gw5n2f": ...  
81740  {"/m/01jfsb": "Thriller", "/m/03npn": "Horror"...  

[81741 rows x 9 columns]

A quick visual insight can be done to get information gain. From the $81,741$ movies, we see that some variables are textual (i.e. `freebase_movieID`, `name`, ...) or numerical (i.e. `box_office_revenue`, `runtime`, ...).

Let's have a more meaningful insight of our current data.

In [3]:
insight(movie)

class  missing_values
wiki_movieID          <class 'int'>            0.00
freebase_movieID      <class 'str'>            0.00
name                  <class 'str'>            0.00
release_date          <class 'str'>            8.44
box_office_revenue  <class 'float'>           89.72
runtime             <class 'float'>           25.02
languages             <class 'str'>            0.00
countries             <class 'str'>            0.00
genres                <class 'str'>            0.00

We see that only few movies ($8,401$) have information about their box office revenue. Also, not all needed attribute are stored in the actual raw data. Since they are important metrics for out analysis, let's try to retrieve as much information as possible via API requests (procedure explained below).

## Clean/Enrich Data

 To enrich our dataset, we use a free to use API ([TMDB](https://www.themoviedb.org/?language=fr)). Some important features are still missing after this procedure. Then, we drop these movies from our analysis.

A specific script has been created to be run once and create our `clean_enrich_movie.csv` dataset. Go to `enrich_data.py` and its linked library `tmdb/tmdbDataLoader.py` for more explanation on how we retrieved these information. Please note that a personal API key is needed to successfully run the script ([create key](https://developer.themoviedb.org/reference/intro/getting-started)). Make sure to create file `.env` with your API bearer token using the `.env_example` as template.

In [4]:
# Load the cleaned and enriched data set (created via enrich_movie_data.py script)
enhanced_movie = pd.read_pickle('dataset/clean_enrich_movies.pickle')

# Display the loaded dataframe
display(enhanced_movie)

box_office_revenue                                           name  \
0            2.782275e+09                                         Avatar   
1            2.185372e+09                                        Titanic   
2            1.511758e+09                                   The Avengers   
3            1.328111e+09  Harry Potter and the Deathly Hallows – Part 2   
4            1.215577e+09                                     Iron Man 3   
...                   ...                                            ...   
11882        1.000000e+00                            Belizaire the Cajun   
11883        1.000000e+00                              A Cry in the Wild   
11884        1.000000e+00                                      Cicak-Man   
11885        1.000000e+00                             Three for the Road   
11886        1.000000e+00                     MVP: Most Valuable Primate   

                                                  genres   tmdb_id  \
0      [Thriller, Science Fiction, Adventure, Compute...   19995.0   
1      [Tragedy, Costume drama, Historical fiction, A...     597.0   
2                              [Science Fiction, Action]   24428.0   
3                   [Drama, Mystery, Fantasy, Adventure]   12445.0   
4          [Science Fiction, Action, Fantasy, Adventure]   68721.0   
...                                                  ...       ...   
11882  [Crime Fiction, Social problem film, Indie, Hi...  274253.0   
11883  [Family Film, Family Drama, Drama, Adventure, ...   25373.0   
11884                          [Comedy, Superhero movie]   69767.0   
11885                             [Drama, Indie, Comedy]  116064.0   
11886  [Children's/Family, Animal Picture, Sports, Ch...   32834.0   

      release_date                                   countries  \
0             2009  [United States of America, United Kingdom]   
1             1997                  [United States of America]   
2             2012                  [United States of America]   
3             2011  [United States of America, United Kingdom]   
4             2013                  [United States of America]   
...            ...                                         ...   
11882         1986                  [United States of America]   
11883         1990                  [United States of America]   
11884         2006                                  [Malaysia]   
11885         1987                  [United States of America]   
11886         2000                  [United States of America]   

                                               composers  
0      [Composer(id=1729, name='James Horner', birthd...  
1      [Composer(id=1729, name='James Horner', birthd...  
2      [Composer(id=37, name='Alan Silvestri', birthd...  
3      [Composer(id=2949, name='Alexandre Desplat', b...  
4      [Composer(id=6041, name='Brian Tyler', birthda...  
...                                                  ...  
11882  [Composer(id=2886167, name='Michael Doucet', b...  
11883  [Composer(id=58124, name='Arthur Kempel', birt...  
11884                                                NaN  
11885  [Composer(id=69900, name='Barry Goldberg', bir...  
11886                                                NaN  

[11887 rows x 7 columns]

Let's have a more deep look at this new dataframe. First, let's have a look of the percentage of missing values now.

In [5]:
insight(enhanced_movie)

class  missing_values
box_office_revenue  <class 'float'>            0.00
name                  <class 'str'>            0.00
genres               <class 'list'>            0.00
tmdb_id             <class 'float'>            0.00
release_date          <class 'str'>            0.00
countries            <class 'list'>            0.00
composers            <class 'list'>           32.99

We also display statistical values to better understand the distribution of few metrics of the data.

In [6]:
insight_enhance(enhanced_movie)

count    1.188700e+04
mean     4.218139e+07
std      1.087840e+08
min      1.000000e+00
25%      1.432743e+06
50%      7.241180e+06
75%      3.200000e+07
max      2.782275e+09
Name: box_office_revenue, dtype: float64

There is 32.99% of nan composers

Considering the first composer of the list if multiple have been returned for a movie, we can compute the following statistics on the retrieved data:

	 - There is 0.00% of nan name for composers
	 - There is 20.39% of nan birthday for composers
	 - There is 0.00% of nan gender for composers
	 - There is 73.24% of nan homepage for composers
	 - There is 24.88% of nan place of birth for composers
	 - There is 0.00% of nan first appearance in movie for composers


## Spotify API

For question 2 and 7 we want to enrich our data with the Spotify API. We want to get the genre and the popularity of each song of each movie. To do so, we use the [Spotify](https://spotipy.readthedocs.io/en/2.16.1/) library. We can then use the `spotify_dataset` to get the genre and the popularity of each song of each composer which have composed at least one music for a movie. As the spotify API has got a limit on the requests number, we were not able to create the complete dataset for now. That's why, we are only showing the data for one composer for this milestone, just to show that we can access the necessary data to answer our research questions.

In [7]:
musics = pd.read_csv('dataset/spotify_dataset.csv')
musics.head()

Unnamed: 0                      id                                name  \
0           0  5XH4I94l2ZyBlA9pB5FAlK  Générique de fin, Dancing in Jaffa   
1           1  3tJUcLIzEW7jXWR6LFFPLj   Les bords du gange, Le grand jour   
2           2  5FPFwymFgzcIvAuzfqdTwx  Générique de début, Les misérables   
3           3  72FY2lq2awOURdaDCtiTaW                 Blanc comme neige I   
4           4  26dHq3RQ5o3mY8md7VXnS7               Le système de Ponzi I   

                   genre             composer_id  popularity  
0  ['french soundtrack']  5qtybJzm3AeeHlABbAaqTC           0  
1  ['french soundtrack']  5qtybJzm3AeeHlABbAaqTC           0  
2  ['french soundtrack']  5qtybJzm3AeeHlABbAaqTC           2  
3  ['french soundtrack']  5qtybJzm3AeeHlABbAaqTC           0  
4  ['french soundtrack']  5qtybJzm3AeeHlABbAaqTC           0